### ISSUE TITLE
Yara search trips over paths using backslashes

### ISSUE THREAD

**Issue by wagnerand** [mozilla/addons-server#18111]
### Describe the problem and steps to reproduce it:

Yara search throws hundreds of errors for every scanner query search, due to some XPIs using backslashes in paths.

### What happened?

Lots of validation errors.

### What did you expect to happen?

No errors.

### Anything else we should know?

https://sentry.prod.mozaws.net/operations/olympia-prod/issues/13707310/

**Comment by bobsilverberg** [comment]
@diox Now that I am pretty much done with https://github.com/mozilla/addons-server/issues/17318, I can start looking at this. Can you give me some pointers? Looking through the related issues, is the problem that the validation at https://github.com/mozilla/addons-server/blob/master/src/olympia/files/utils.py#L559 is being triggered, or is it more than that? Where does the code for "Yara search" live?

**Comment by diox** [comment]
Yara searches are triggered by going to the admin for scanners, creating a new scanner query rule (the "query" part is the important bit). Then, trigerring a rule will fire a group of tasks that perform a search for the yara query you entered across all known xpis from your database (*). You can then view the results through the admin - there should be a link to the corresponding scanner query results.

When performing the search, we re-use the `SafeZip` class to open the `xpis` while being protected from zipbombs etc. As noted above, this now triggers the new checks we added for paths inside the zip that contain backslashes.

(*) Which means to test it, you need a database that has corresponding files on your filesystem. You might need to upload a couple test add-ons.

**Comment by bobsilverberg** [comment]
Thanks @diox. Based on this, I am guessing the line causing the exception is https://github.com/mozilla/addons-server/blob/master/src/olympia/scanners/tasks.py#L251, correct?

**Comment by diox** [comment]
Yes. I'm not 100% sure what we should do about that. Maybe a way to call this that bypasses this specific check so that we can open the zip and run the search on it or catch the error but ignore that file (dangerous, would need to ensure the corresponding versions are all blocked first), or something else entirely...

**Comment by bobsilverberg** [comment]
Some basic testing suggests that if we bypass the checks for invalid filenames (https://github.com/mozilla/addons-server/blob/master/src/olympia/files/utils.py#L558-L567) when we are running a code search, it will still search through the files and find matches in any files that *do* have valid filenames. I even tested with an invalid file in a folder beside a file that should be found, and it was able to scan the neighbour file. Would this be a reasonable solution, or might there be other problems related to bypassing those checks when doing a code search?

**Comment by diox** [comment]
That sounds promising. I'm wondering though... those name checks are mostly there to avoid issues at extraction or in Firefox, but our scanners python code shouldn't care, right ? It'd be nice if we bypassed that particular check about filenames but were still able to find matches against those files.

**Comment by AlexandraMoga** [comment]
I've run a quick regression on scanner query results on stage by searching for some specific matched rules in various add-ons. I've also included add-ons that were force disabled or deleted. I haven't noticed any issues with the results I've got.
This are the results for one scanner query rule I've run: https://addons-internal.stage.mozaws.net/en-US/admin/models/scanners/scannerqueryresult/?matched_rules__id__exact=12

For the verification of the actual issue that was mentioned in the STR, we'll have to wait for this patch to land in prod and run some scanner query rules there. Submission of files using backslashes is not possible anymore, so we did not have the required test data on -dev and stage to be able to verify the fix proper.
